lib/src/embedded/compiler.ts

```diff
@@ -118,10 +84,132 @@ export class EmbeddedCompiler {
       .subscribe(() => process.nextTick(() => this.close()));
   }
 
+  /**
+   * Dispatches a compile request. Returns a promise that resolves when a
+   * compile response with a matching ID is received from the embedded compiler.
+   */
+  compile(request: InboundMessage.CompileRequest) {
+    const id = this.nextRequestId();
+    request.setId(id);
+    const message = new InboundMessage();
+    message.setCompilerequest(request);
+
+    this.recordRequest(id);
+    this.messageTransformer.write$.next(message);
+
+    return this.messageTransformer.read$
+      .pipe(
+        filter(message => {
+          return (
+            message.hasCompileresponse() &&
+            message.getCompileresponse()!.getId() === id
+          );
+        }),
+        map(message => message.getCompileresponse()!),
+        tap(response => this.recordResponse(response.getId())),
+        take(1),
+        takeUntil(this.error$)
+      )
+      .toPromise();
+  }
+
+  listen(
+    messageType: OutboundMessage.MessageCase,
+    processRequest: (request: Request) => Response
+  ) {
```

Comment:
Rather than having a generic public `listen()` method, I think it would be better to expose separate methods for the various types of requests that can be listened for, such as:

```ts
onImportRequest(
  (request: OutboundMessage.ImportRequest) => Promise<InboundMessage.ImportResponse>
): Subscription
```

That way, you have a type-system-level guarantee that the caller is not only returning a response, but returning the appropriate response type.

(I also think you'll want to make the handler asynchronous since you'll potentially be calling out to async importer and function implementations.)

Reply:
Thanks, that's a better solution.